Fix for Tracing Models with Backpropagation

oldio opened the pull request:
**Fix for Tracing Models with Backpropagation (#120820)**

***Issue Description***
This PR addresses critical issues encountered when tracing models involving backpropagation. The current implementation fails for seemingly simple scenarios, such as:
```
import torch

class Model(torch.nn.Module):
    def forward(self, inputs):
        x = inputs[0].detach().requires_grad_()
        with torch.enable_grad():
            loss = torch.sqrt(x).sum()
            return torch.autograd.grad([loss], [x])[0]

torch.jit.trace(Model(), torch.FloatTensor(2, 3, 4).uniform_(0, 10))
```

This results in a misleading error as no constants are used during the model computation:
```
RuntimeError: Cannot insert a Tensor that requires grad as a constant. Consider making it a parameter or input, or detaching the gradient
```


***Root Cause Analysis***
The problem stems from the `TracingState`'s use of `IWeakValue` to `Value` mappings:
```
using Frame = std::unordered_map<WeakIValue, Value*, WeakIValueHasher, WeakIValueEq>;
std::vector<Frame> env_stack;
```

Tensors use the `TensorImpl` address as a hash. However, backward function implementations (e.g., `SqrtBackward0::apply()`) create new `TensorImpl` instances when unpacking `SavedVariable` members via `make_variable`:
```
// in SavedVariable::unpack() function...
Variable var;
if (grad_fn) {
    var = make_variable(data, Edge(std::move(grad_fn), output_nr_));
} else { ...
```
These new instances aren't pre-registered in the tracing state `env_stack`, leading to lookup failures and the observed error.


***Solution***
To resolve this, we've implemented a two-step lookup process in the `TracingState`:

1. First, attempt to find the `TensorImpl` using the existing hash-based lookup.
2. If not found, perform a secondary linear search to find a matching `TensorImpl` based on storage alias and tensor metadata equality (e.g. storage alias and a copy of metadata properties are both performed by `make_variable`)

This approach ensures compatibility with tensors created during backpropagation, allowing successful tracing of models with gradient computation.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @thiagocrepaldi - I'm bringing this PR to your attention given you discussed this particular issue with my colleague @DavidMinorNV in the past. Feel free to add any reviewers for feedback.

oldio committed:
Fix pytorch jit-trace to consider temporary tensors created when calling make_variable from SavedVariable::unpack()

oldio committed:
Disable debug code and fix comments

oldio committed:
Added simple tracing tests with autograd

oldio commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -1429,6 +1429,29 @@ struct TORCH_API WeakIValue final {
         is_intrusive_ptr == rhs.is_intrusive_ptr;
   }
 
+#if 0
```
Left this debug code temporarily for review purposes.

oldio commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -166,6 +178,21 @@ Value* TracingState::getValue(const IValue& var) {
 
     // Didn't find it. Bake in a constant
     if (ten.requires_grad()) {
+#if 0
```
Left this debug code temporarily for review purposes.

oldio committed:
Fix comments

bdhirsh commented:
maybe cc @davidberard98 for jit.trace ?

davidberard98 commented:
@pytorchbot rebase -s

davidberard98 commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -153,7 +153,19 @@ Value* TracingState::getValue(const IValue& var) {
-        continue;
+        // If we can't find corresponding TensorImpl based on ptr hash, perform search by
+        // trying to find a tensor based on matching storage alias and metadata...
+        it = value_map.begin();
```
to limit the number of affected scenarios, is it possible to only take this path if the tensor requires grad? or are there cases where we have saved tensors that don't require grad?

oldio replied:
Good point @davidberard98! Let me take a look as I'm pretty sure we should be able to isolate it.

oldio replied:
Isolated the linear search only if tensor requires a gradient (https://github.com/pytorch/pytorch/pull/134879/commits/686110b0ecf750ad329a6ef3ff465aa846dc427e)

davidberard98 reviewed (commented):
Thanks for the detailed PR description, it helps a lot with the review!

The main thing I'm worried about is whether there are real world cases where the change in tracer.cpp could affect the behavior of the traced program. I couldn't come up with very good examples, but the example below does differ in its output when compared to before applying this PR:

```python
import torch

x = torch.ones((4,))
y = x.detach()
z = torch.ones((4,))

def fn(z):
    global x
    a = x * z
    x += 1
    b = y * z
    return a + b

fn_t = torch.jit.trace(fn, (z,))

print(fn_t(z))
```

anyway, for now I've rebased the PR on viable/strict (base branch on which all the tests should pass) to see if we can get any CI signal. I'll take another look tomorrow (feel free to ping me if I forget)

davidberard98 commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -553,6 +580,11 @@ void TracingState::setValue(const IValue& v, Value* value) {
+    // Hold tensor reference for the duration of tracing so that
+    // temporary Variables (i.e. created in SavedVariable::unpack when calling make_variable)
+    // are not released too early
+    backup_.push_back(var);
```
actually more concerned about this:

```python
import torch

def fn(z):
    r = z
    for i in range(200):
        r += torch.ones(2**30, device="cuda")
    return r

z = torch.ones(2**30, device="cuda")

fn_t = torch.jit.trace(fn, (z,))

print(fn_t(z).sum())
```

before this change, this works without problems; after this, it runs out of memory

oldio replied:
Thanks @davidberard98, I'll take a look (and also add as a test case)...  we can probably limit holding references to aliased tensors created by make_variable. I'll investigate.

oldio replied:
Isolated backup to tensors that require a gradient (https://github.com/pytorch/pytorch/pull/134879/commits/686110b0ecf750ad329a6ef3ff465aa846dc427e)

Dasyud replied:
Hi @oldio,
Thanks a lot for this contribution! I have been looking into the problem for a while now since I have a similar use case. If I want to start working with your changes ASAP, do I need to build pytorch from source with your commits?

oldio replied:
Cheers @Dasyud and yes you'd need to build from source given some of the tracing backend cpp files have been modified.

Dasyud replied:
Hi @oldio ,
I built pytorch with your PR and ran my code again, but the error hasn't changed (cannot insert constant). Could you please try running my example code in your environment? Maybe there is a bug in my code.

https://github.com/pytorch/pytorch/issues/125984#issuecomment-2348152846

oldio replied:
The problem you linked seems to stem from another issue outside the scope of this current PR. Notice that the error message is not the same... in your case it's:
`RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn`
instead of:
`RuntimeError: Cannot insert a Tensor that requires grad as a constant. Consider making it a parameter or input, or detaching the gradient`

Your example is also using dynamo to export rather then torch.jit.trace (different backends as I understand) but nevertheless I did run a few other failing cases with my branch and issue remains.

Dasyud replied:
i'm sorry, there's a slight change in the code that i failed to include in the link i shared.
i realised that too, so i passed enable_grad=True to self.qf.

```
    def calc_q_grad(self, state, action):
        qf_output = self.qf(state, action, enable_grad=True)
```

Dasyud replied:
Hi @oldio ,
Did you get a chance to run the code again?

Dasyud replied:
I tested it again, and noticed there has been a difference. The error has changed from FloatTensor{1, 1} to FloatTensor{128, 1}.
There probably is another blocker in my code that is not letting me export.

oldio commented:
> Thanks for the detailed PR description, it helps a lot with the review!
> 
> The main thing I'm worried about is whether there are real world cases where the change in tracer.cpp could affect the behavior of the traced program. I couldn't come up with very good examples, but the example below does differ in its output when compared to before applying this PR:
> 
> ```python
> import torch
> 
> x = torch.ones((4,))
> y = x.detach()
> z = torch.ones((4,))
> 
> def fn(z):
>     global x
>     a = x * z
>     x += 1
>     b = y * z
>     return a + b
> 
> fn_t = torch.jit.trace(fn, (z,))
> 
> print(fn_t(z))
> ```
> 
> anyway, for now I've rebased the PR on viable/strict (base branch on which all the tests should pass) to see if we can get any CI signal. I'll take another look tomorrow (feel free to ping me if I forget)

The tr

oldio commented:
Sorry all closed the PR by mistake (sigh...)

oldio committed:
Address PR comments

oldio commented:
@davidberard98 I did another pass on the code (https://github.com/pytorch/pytorch/pull/134879/commits/686110b0ecf750ad329a6ef3ff465aa846dc427e) based on your PR comments

oldio commented on test/jit/test_trace_autograd.py:
```diff
@@ -0,0 +1,70 @@
+    def validateTrace(self, model):
+        inputs = torch.FloatTensor(2, 3, 4).uniform_(0, 10)
+        traced = torch.jit.trace(model, inputs)
+        self.assertEqual(abs(traced(inputs) - model(inputs)).sum(), 0.)
```
Compare that traced model vs original give the same expected results

oldio committed:
Run and apply lintrunner

oldio commented:
Hi @bdhirsh and @davidberard98, just wanted to follow up on this PR and am wondering how you'd like to proceed. I addressed comments and failing lint tests last week let me know if you'd like me to do any more changes.

oldio commented:
cc @colesbury for visibility on [comment](https://github.com/pytorch/pytorch/pull/134879#issuecomment-2352800257)

oldio committed:
Fix lint failure test_jit.py

Dasyud commented:
Just listing down an example where the error does not change.

```
import torch
import torch.nn as nn
import numpy as np
from torch.distributions import MultivariateNormal

LOG_STD_MIN = -5.0
LOG_STD_MAX = 2.0

DEFAULT_DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
# DEFAULT_DEVICE = torch.device('cpu')

def torchify(x):
    x = torch.from_numpy(x)
    if x.dtype is torch.float64:
        x = x.float()
    x = x.to(device=DEFAULT_DEVICE)
    return x

class Squeeze(nn.Module):
    def __init__(self, dim=None):
        super().__init__()
        self.dim = dim

    def forward(self, x):
        return x.squeeze(dim=self.dim)
    
def mlp(dims, activation=nn.ReLU, output_activation=None, squeeze_output=False):
    n_dims = len(dims)
    assert n_dims >= 2, 'MLP requires at least two dims (input and output)'

    layers = []
    for i in range(n_dims - 2):
        layers.append(nn.Linear(dims[i], dims[i+1]))
        layers.append(activation())
    layers.append(nn.Linear(dims[-2], dims[-1]))
    if output_activation is not None:
        layers.append(output_activation())
    if squeeze_output:
        assert dims[-1] == 1
        layers.append(Squeeze(-1))
    net = nn.Sequential(*layers)
    net.to(dtype=torch.float32)
    return net

# had to define this because torch.diag_embed was not being onnx exported
def custom_diag_embed(diagonal):
    batch_size, dim = diagonal.shape
    return torch.eye(dim).unsqueeze(0).repeat(batch_size, 1, 1).to(diagonal.device) * diagonal.unsqueeze(-1)

class GaussianPolicy(nn.Module):
    def __init__(self, obs_dim, act_dim, hidden_dim=256, n_hidden=2, state_min = None, state_max = None, state_mean = None, state_std = None):
        super().__init__()
        self.net = mlp([obs_dim, *([hidden_dim] * n_hidden), act_dim], output_activation=nn.Tanh)
        self.state_min = state_min
        self.state_max = state_max
        self.state_mean = state_mean
        self.state_std = state_std

    def forward(self, obs):
        if self.state_mean is not None:
            obs = (obs - self.state_mean) / self.state_std
        mean, log_std = torch.chunk(self.net(obs), 2, dim=-1)
        std = torch.exp(log_std.clamp(LOG_STD_MIN, LOG_STD_MAX))
        # scale_tril = torch.diag(std)
        # scale_tril = torch.diag_embed(std)
        scale_tril = custom_diag_embed(std)
        return MultivariateNormal(mean, scale_tril=scale_tril)

    def act(self, obs, deterministic=True, enable_grad=False): # only called during evaluation
        with torch.set_grad_enabled(enable_grad):
            dist = self(obs) # dist is a MultivariateNormal object, calls forward
            return dist.mean, dist.stddev if deterministic else dist.sample()

class TwinQ(nn.Module):
    def __init__(self, state_dim, action_dim, hidden_dim=256, n_hidden=2, state_min = None, state_max = None, state_mean = None, state_std = None):
        super().__init__()
        dims = [state_dim + action_dim, *([hidden_dim] * n_hidden), 1]
        self.q1 = mlp(dims, activation=nn.Tanh, squeeze_output=True)
        self.q2 = mlp(dims, activation=nn.Tanh, squeeze_output=True)
        self.state_min = state_min
        self.state_max = state_max
        self.state_mean = state_mean
        self.state_std = state_std

    def both(self, state, action):
        sa = torch.cat([state, action], 1)
        # sa = sa.float() # added this line
        return self.q1(sa), self.q2(sa)

    def forward(self, state, action, enable_grad=False):
        with torch.set_grad_enabled(enable_grad):
            if self.state_mean is not None:
                state = (state - self.state_mean) / self.state_std
            return torch.min(*self.both(state, action))

class QFGrad(nn.Module):
    def __init__(self, policy, qf, beta=0):
        super().__init__()
        self.policy = policy
        self.qf = qf
        self.beta = beta

    def forward(self, state, hidden_state, cell_state):
        state = state.squeeze(1)
        action, _ = self.policy.act(state, enable_grad=True)        
        updated_action = action + self.beta * self.calc_q_grad(state, action)
        return updated_action, hidden_state, cell_state

    def calc_q_grad(self, state, action):
        qf_output = self.qf(state, action, enable_grad=True)
        # return torch.autograd.functional.jacobian(self.qf, (state, action), create_graph=False)[1]
        return torch.autograd.grad(qf_output, action, create_graph=True, retain_graph=True)[0]

def main():

    obs_dim = 150
    act_dim = 1

    policy = GaussianPolicy(obs_dim, act_dim * 2, hidden_dim=128, n_hidden=2)
    qf = TwinQ(obs_dim, act_dim, hidden_dim=128, n_hidden=2)

    policy.to(DEFAULT_DEVICE)
    qf.to(DEFAULT_DEVICE)

    # create a meta model with policy and qf, that follows the equation: output = policy(input) + grad_a * qf(input, policy(input))
    meta_model = QFGrad(policy, qf, beta=0.05)

    # Example usage
    # input_tensor = torch.from_numpy(np.random.rand(1, 1, 150)).float()
    input_tensor = torchify(np.random.rand(1, 1, 150))
    dummy_hidden = torch.zeros(1, 1)
    dummy_cell = torch.zeros(1, 1)
    
    # Export to ONNX with additional inputs and outputs
    torch.onnx.export(meta_model, 
                  (input_tensor, dummy_hidden, dummy_cell),
                  "demo_1.onnx", 
                  input_names=["obs", "hidden_states", "cell_states"],
                  output_names=["output", "state_out", "cell_out"],)
    
if __name__ == '__main__':
    main()
```
